tests/cbmc/source/proof_allocators.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <cbmc_proof/proof_allocators.h>
+#include <stdarg.h>
+#include <stdlib.h>
+
+void *bounded_calloc(size_t num, size_t size) {
+    size_t required_bytes;
+    __CPROVER_assume(!__builtin_mul_overflow(num, size, &required_bytes));
+    __CPROVER_assume(required_bytes <= MAX_MALLOC);
+    return calloc(num, size);
+}
+
+void *bounded_malloc(size_t size) {
+    __CPROVER_assume(size <= MAX_MALLOC);
```

Comment:
This worries me a bit, because anybody that calls `bounded_malloc(MAX_MALLOC + 1)` invalidates your entire proof. Is this function something you always call? If not, could you assert this condition instead of assuming it?

Reply:
The function `can_fail_malloc` returns `NULL` instead of assuming in this case.  I could add a third case which asserts that the input is smaller.  But the standard way of using this is 

```
size_t len;
char* mem = bounded_malloc(len);
```

In which case, we want to test for all sizes which CBMC allows.